src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,406 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class LocalRewriter
+    {
+        /// <summary>
+        /// Rewrite `GetTuple() == (1, 2)` to `tuple.Item1 == 1 &amp;&amp; tuple.Item2 == 2`.
+        /// Also supports the != operator, nullable and nested tuples.
+        /// Note that all the side-effects for visible expressions are evaluated first (into temporaries as needed).
+        /// Element-wise conversions occur late, together with the element-wise comparisons. They may not be evaluated.
+        /// </summary>
+        /// There are 2 cases:
+        /// - tuple literal cannot be null and it produces as many temps as elements
+        /// - tuple types or nullable tuple types produce a single temp (and no further temps are initialized)
+        public override BoundNode VisitTupleBinaryOperator(BoundTupleBinaryOperator node)
+        {
+            var boolType = node.Type; // we can re-use the bool type
+            var initialEffectsAndTemps = TupleOperatorSideEffectsAndTemps.GetInstance();
+
+            var returnValue = RewriteTupleOperator(node.Operators, node.Left, node.Right, boolType, initialEffectsAndTemps, node.OperatorKind, inLeftInit: true, inRightInit: true);
+
+            var (effects, temps) = initialEffectsAndTemps.ToImmutableAndFree();
+            var result = _factory.Sequence(temps, effects, returnValue);
+            return result;
+        }
+
+        private BoundExpression RewriteTupleOperator(TupleBinaryOperatorInfo @operator,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType,
+            TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftInit, bool inRightInit)
```

Comment:
What do `inLeftInit` and `inRightInit` represent?

Reply:
In `(a, b, tuple) == (c, d, (e, f))` all the visible elements (a, b, tuple, c, d, e, and f) are considered "in init". But `tuple.Item1` and `tuple.Item2` are not (they will not necessarily be evaluated, especially if `tuple` is actually a nullable type).
When "in init" we'll evaluate the tuple elements into the init sequence which is always evaluated. When drilling into an expression that isn't a tuple literal (ie we're no longer "in init"), the evaluation becomes conditional as that element is only meaningful in a certain scope/condition.

I have a comment at the top that hints this, but I agree it's not fully explained: `/// Note that all the side-effects for visible expressions are evaluated first (into temporaries as needed).`


---
In reply to: [168004240](https://github.com/dotnet/roslyn/pull/23356#discussion_r168004240) [](ancestors = 168004240)